source-academy/frontend
# Add Interactive Environment Visualiser

### Description
#### Dependencies
This version of js-slang https://github.com/source-academy/js-slang/pull/1403 is needed for this implementation to work correctly.

#### Motivation 
The current implementation of the environment visualizer only allows users to see the state of the environment at different stages of the program by adding breakpoints at every stage where they want to see the environment model.

#### Summary
This pull request adds a slider to the environment visualizer which allows users to interactively step through and see how the environment changes at different stages of their program. This is achieved by initially evaluating the program all the way to the end to count the number of steps, and then subsequently partially evaluating the program until the step number is requested by the user. The program is reevaluated from the start if the user edits their program.

Currently, the slider evaluates on Release rather than on Change due to efficiency reasons since the current implementation reruns the program when the slider is shifted. In the future, this would likely be changed to an onChange mechanism by finding a more efficient way instead of rerunning the program from the start.

Moreover, a breakpoint is no longer required in order for the environment visualizer to be activated, the users can simply activate this mode by switching to the tab. Breakpoints now instead of breaking the program, allow the users to fast-forward their slider to the breakpoint. This allows for the old functionality of the environment visualizer to also be achieved where users could choose to see the state of the environment model right before a specific line is executed.



### Type of change


### How to test


1. Run your local version of Source Academy linked to your own local js-slang. The local js-slang version must be from https://github.com/source-academy/js-slang/pull/1403.
2. Enter a program in the playground, and switch to the environment visualizer tab.
3. Run the program and step through to see the changes.
4. The program can be edited and then rerun to see the updated environment.
Note: Not compatible with modules.

### Checklist
